### PR TITLE
User login

### PR DESCRIPTION
**Pull request by jersonlopez**
## Summary

login an user and merge with HEAD

## Trello Card

[https://trello.com/c/NSGh1otZ/3-sign-in](https://trello.com/c/NSGh1otZ/3-sign-in)

### CODE REVIEW

File: app/controllers/user.js [hunk 1]
```diff
@@ -5,7 +5,10 @@ const bcrypt = require('bcrypt'),
   salt = bcrypt.genSaltSync(saltRounds),
   errors = require('../errors'),
   { validateUser } = require('./validations'),
-  User = require('../models').users;
+  User = require('../models').users,
+  jwt = require('jwt-simple'),
+  logger = require('../logger'),
+  secret = 'super-secret';
```
Comment (picollomartin) [inline app/controllers/user.js]: This should be an environment variable, not hardcoded

File: app/models/user.js [hunk 2]
```diff
@@ -45,5 +45,16 @@ module.exports = (sequelize, DataTypes) => {
         }
         throw errors.databaseError(err);
       });
+
+  User.getUserByEmailAndPassword = email =>
```
Comment (picollomartin) [inline app/models/user.js]: `getUserByEmailAndPassword` weird name for function, you can do more generic like `getUserBy = condition => ....`

File: app/controllers/user.js [hunk 3]
```diff
@@ -34,3 +37,41 @@ exports.singUp = (req, res, next) => {
...
+    : {};
+
+  const signErrors = validateUser(user);
+
+  try {
+    if (!signErrors.valid) throw errors.signInError(signErrors.messages);
+
+    return User.getUserByEmailAndPassword(user.email)
+      .then(result => {
+        if (!result) throw errors.signInError(['user not registered']);
```
Comment (picollomartin) [inline app/controllers/user.js]: Send directly the message not as an array (check this in all errors in this function)

File: app/controllers/user.js [hunk 4]
```diff
@@ -34,3 +37,41 @@ exports.singUp = (req, res, next) => {
...
+
+        bcrypt.compare(user.password, result.password, (err, validPassword) => {
+          if (validPassword) {
+            logger.info(`${result.firstName} logged in.`);
+            const token = jwt.encode({ email: result.email }, secret);
```
Comment (picollomartin) [inline app/controllers/user.js]: I think in a service for handle all authentication functions, what do you think ?

File: app/controllers/user.js [hunk 5]
```diff
@@ -34,3 +37,41 @@ exports.singUp = (req, res, next) => {
...
+            res
+              .status(200)
+              .set('auth', token)
```
Comment (picollomartin) [inline app/controllers/user.js]: Use a common header for authentication like `Authorization` (put this in a environment variable)

File: app/controllers/user.js [hunk 6]
```diff
@@ -34,3 +37,41 @@ exports.singUp = (req, res, next) => {
...
+  const signErrors = validateUser(user);
+
+  try {
+    if (!signErrors.valid) throw errors.signInError(signErrors.messages);
+
+    return User.getUserByEmailAndPassword(user.email)
+      .then(result => {
+        if (!result) throw errors.signInError(['user not registered']);
+
+        bcrypt.compare(user.password, result.password, (err, validPassword) => {
```
Comment (picollomartin) [inline app/controllers/user.js]: Return this promise too

File: test/user.js [hunk 7]
```diff
@@ -127,4 +127,116 @@ describe('users', () => {
...
+
+  describe('/users/sessions/ POST', () => {
+    it('should login an user without problems', done => {
+      chai
+        .request(server)
+        .post('/users/')
+        .send({
+          firstName: 'pepito',
+          lastName: 'perez',
+          email: 'pepito.perez@wolox.com',
```
Comment (picollomartin) [inline test/user.js]: Make a function for create users, check this in all test file

File: app/middlewares/errors.js [hunk 8]
```diff
@@ -10,6 +10,7 @@ const statusCodes = {
   [errors.DATABASE_ERROR]: 503,
   [errors.SAVING_ERROR]: 400,
   [errors.SIGNUP_ERROR]: 400,
+  [errors.SIGNIN_ERROR]: 404,
```
Comment (picollomartin) [inline app/middlewares/errors.js]: More like 400 code too, 404 is not found

Comment (picollomartin) [inline app/middlewares/errors.js]: More like 400 code too, 404 is not found

File: app/controllers/user.js [hunk 9]
```diff
@@ -34,3 +37,41 @@ exports.singUp = (req, res, next) => {
...
+              .set('auth', token)
+              .send(result);
+            res.end();
+          } else {
+            next(errors.signInError(['Password invalid!']));
+          }
+        });
+      })
+      .catch(next);
+  } catch (err) {
```
Comment (picollomartin) [inline app/controllers/user.js:70]: Use promise way or try-catch with await, both together are too verbose

File: app/controllers/user.js [hunk 10]
```diff
@@ -34,3 +37,41 @@ exports.singUp = (req, res, next) => {
...
+    next(err);
+  }
+  res.send(req.body);
```
Comment (picollomartin) [inline app/controllers/user.js]: Remove this

**Review by picollomartin — CHANGES_REQUESTED** [review]

File: test/user.js (hunk 7 above)
Comment (jersonlopez) [reply test/user.js]: like as?

File: app/controllers/user.js (hunk 6 above)
Comment (jersonlopez) [reply app/controllers/user.js]: why? if it is working here

File: app/controllers/user.js [hunk 11]
```diff
@@ -5,7 +5,9 @@ const bcrypt = require('bcrypt'),
   salt = bcrypt.genSaltSync(saltRounds),
   errors = require('../errors'),
   { validateUser } = require('./validations'),
-  User = require('../models').users;
+  User = require('../models').users,
+  { encoder, HEADER } = require('./../services/jwt'),
```
Comment (picollomartin) [inline app/controllers/user.js:9]: The service name could be not coupled to implementation, you can name this like `authentication` or something better

**Review by picollomartin — APPROVED** [review]
Nice job ! :tada:

**Pull request merged**